janusgraph-driver/src/main/java/org/janusgraph/core/attribute/Text.java

```diff
@@ -236,16 +533,168 @@ public boolean evaluateRaw(String value, String term) {
+            this.preevaluate(value, condition);
+            return value != null && evaluateRaw(value.toString(), (String) condition);
+        }
+
+        @Override
+        public boolean evaluateRaw(String value, String terms) {
+            List<String> valueTerms = tokenize(value.trim().toLowerCase());
+            List<String> tokenTerms = tokenize(terms.trim().toLowerCase());
+            if (!terms.isEmpty() && tokenTerms.isEmpty()) return false;
+            return (Collections.indexOfSubList(valueTerms, tokenTerms) != -1);
+        }
+
+        @Override
+        public boolean isValidCondition(Object condition) {
+            return condition instanceof String && StringUtils.isNotBlank((String) condition);
+        }
+
+        @Override
+        public boolean hasNegation() {
+            return true;
+        }
+
+        @Override
+        public JanusGraphPredicate negate() {
+            return Text.NOT_CONTAINS_PHRASE;
+        }
+
+        @Override
+        public String toString() {
+            return "textContainsPhrase";
+        }
+    },
+
+    /**
+     * Whether the text does not contains a given token sequence in the text (case insensitive)
+     */
+    NOT_CONTAINS_PHRASE {
+
+        @Override
+        public boolean test(Object value, Object condition) {
+            this.preevaluate(value, condition);
+            return value != null && evaluateRaw(value.toString(), (String) condition);
+        }
+
+        @Override
+        public boolean evaluateRaw(String value, String terms) {
+            List<String> valueTerms = tokenize(value.trim().toLowerCase());
+            List<String> tokenTerms = tokenize(terms.trim().toLowerCase());
+            if (!terms.isEmpty() && tokenTerms.isEmpty()) return true;
+            return (Collections.indexOfSubList(valueTerms, tokenTerms) == -1);
+        }
+
+        @Override
+        public boolean isValidCondition(Object condition) {
+            return condition instanceof String && StringUtils.isNotBlank((String) condition);
+        }
+
+        @Override
+        public boolean hasNegation() {
+            return true;
+        }
+
+        @Override
+        public JanusGraphPredicate negate() {
+            return Text.CONTAINS_PHRASE;
+        }
+
+        @Override
+        public String toString() {
+            return "textNotContainsPhrase";
+        }
     };
 
     private static final LevenshteinDistance ONE_LEVENSHTEIN_DISTANCE = new LevenshteinDistance(1);
     private static final LevenshteinDistance TWO_LEVENSHTEIN_DISTANCE = new LevenshteinDistance(2);
 
+    /**
+     * Calculates the max fuzzy edit distance for a term given its length
+     * - 0 for strings of one or two characters
+     * - 1 for strings of three, four or five characters
+     * - 2 for strings of more than five characters
+     * @param term
+     * @return
+     */
+    public static int getMaxEditDistance(String term) {
```

Comment:
nitpick: Can you move this method before `private static final LevenshteinDistance ONE_LEVENSHTEIN_DISTANCE = new LevenshteinDistance(1);` because `ONE_LEVENSHTEIN_DISTANCE` is only needed by the method `isFuzzy` below.